src/core/include/iDynTree/Core/MatrixView.h

```diff
@@ -0,0 +1,279 @@
+        {
+        }
+
+        template <
+            class Container,
+            std::enable_if_t<std::is_const<element_type>::value
+                                 && std::is_convertible<decltype(std::declval<Container>().data()),
+                                                        pointer>::value
+                                 && !has_IsRowMajor<Container>::value
+                                 && !std::is_same<Container, MatrixView>::value,
+                             int> = 0>
+        MatrixView(const Container& matrix, const StorageOrder& order = StorageOrder::RowMajor)
+            : MatrixView(matrix.data(), matrix.rows(), matrix.cols(), order)
+        {
+        }
+
+        template <class Container,
+                  std::enable_if_t<
+                      std::is_convertible<decltype(std::declval<Container>().data()), pointer>::value
+                          && has_IsRowMajor<Container>::value
+                          && !std::is_same<Container, MatrixView>::value,
+                      int> = 0>
+        MatrixView(Container& matrix)
+            : MatrixView(matrix.data(),
+                         matrix.rows(),
+                         matrix.cols(),
+                         Container::IsRowMajor ? StorageOrder::RowMajor
+                                               : StorageOrder::ColMajor)
+        {
+        }
+
+        template <class Container,
+                  std::enable_if_t<
+                      std::is_convertible<decltype(std::declval<Container>().data()), pointer>::value
+                          && !has_IsRowMajor<Container>::value
+                          && !std::is_same<Container, MatrixView>::value,
+                      int> = 0>
+        MatrixView(Container& matrix, const StorageOrder& order = StorageOrder::RowMajor)
+            : MatrixView(matrix.data(), matrix.rows(), matrix.cols(), order)
+        {
+        }
+
+        MatrixView(pointer in_data,
+                   const unsigned int& in_rows,
+                   const unsigned int& in_cols,
+                   const StorageOrder& order = StorageOrder::RowMajor)
+            : m_storage(in_data)
+            , m_rows(in_rows)
+            , m_cols(in_cols)
+            , m_storageOrder(order)
+        {
+        }
+
+        const StorageOrder& storageOrder() const noexcept
+        {
+            return m_storageOrder;
+        }
+
+        pointer data() const noexcept
+        {
+            return m_storage;
+        }
+
+        /**
+         * @name Matrix interface methods.
+         * Methods exposing a matrix-like interface to MatrixView.
+         *
+         */
+        ///@{
+        reference operator()(const unsigned int row, const unsigned int col) const
+        {
+            assert(row < this->rows());
+            assert(col < this->cols());
+            return this->m_storage[rawIndex(row, col)];
+        }
+
+        unsigned int rows() const
```

Comment:
Maybe the size type can also be specified as the value type.

Reply:
In case of `span` we have
```cpp
using index_type = std::ptrdiff_t;
```

Is it ok using this?

Reply:
Yes, that is simply ``size_t``. It is just to have the same rationale of ``pointer`` and ``reference``.

Reply:
Done here https://github.com/robotology/idyntree/pull/734/commits/67f25899a801ccdffe0d33506af6d185ae15ff4f